db/records/operations/group.py

```diff
@@ -1,123 +1,240 @@
-        limit:    int, gives number of rows to return
-        offset:   int, gives number of rows to skip
-
-    If no table is given, the column_list must consist entirely of
-    SQLAlchemy column objects associated with a table.
+        table:      SQLAlchemy table object
+        group_by:   GroupBy object giving args for grouping
     """
-    if table is not None:
-        column_objects = create_col_objects(table, column_list)
+    grouping_columns = group_by.get_validated_group_by_columns(table)
+
+    if group_by.mode == GroupMode.PERCENTILE.value:
+        query = _get_percentile_range_group_select(
+            table, grouping_columns, group_by.num_groups
+        )
+    elif group_by.mode == GroupMode.DISTINCT.value:
+        query = _get_distinct_group_select(table, grouping_columns)
     else:
-        column_objects = column_list
-    try:
-        assert all([type(col) == Column for col in column_objects])
-    except AssertionError as e:
-        raise e
-
-    query = (
-        select(*column_objects)
-        .distinct()
-        .limit(limit)
-        .offset(offset)
+        raise rec_exc.BadGroupFormat("Unknown error")
+    return query
+
+
+def _get_distinct_group_select(table, grouping_columns):
+    window_def = GroupingWindowDefinition(
+        order_by=grouping_columns, partition_by=grouping_columns
     )
-    result = execute_query(engine, query)
-    if output_table is not None:
-        column_objects = [output_table.columns[col.name] for col in column_objects]
-    return [tuple(zip(column_objects, row)) for row in result]
 
+    group_id_expr = func.dense_rank().over(
+        order_by=window_def.order_by, range_=window_def.range_
+    )
+    return select(
+        table,
+        _get_group_metadata_definition(window_def, grouping_columns, group_id_expr)
+    )
 
-def _get_filtered_group_by_count_query(
-        table, engine, group_by, limit, offset, order_by, filters, count_query
-):
-    # Get the list of groups that we should count.
-    # We're considering limit and offset here so that we only count relevant groups
-    relevant_subtable_query = get_query(table, limit, offset, order_by, filters)
-    relevant_subtable_cte = relevant_subtable_query.cte()
-    cte_columns = create_col_objects(relevant_subtable_cte, group_by)
-    distinct_tuples = get_distinct_tuple_values(cte_columns, engine, output_table=table)
-    if distinct_tuples:
-        limited_filters = [
-            {
-                "or": [
-                    append_distinct_tuples_to_filter(distinct_tuple_spec)
-                    for distinct_tuple_spec in distinct_tuples
-                ]
-            }
-        ]
-        filtered_count_query = apply_filters(count_query, limited_filters)
-    else:
-        filtered_count_query = None
-    return filtered_count_query
 
+def _get_percentile_range_group_select(table, columns, num_groups):
+    column_names = [col.name for col in columns]
+    CUME_DIST = 'cume_dist'
```

Comment:
I think it would be clearer to have more verbose variable names here (I assume this is "cumulative distribution")

Reply:
In this case, it actually _is_ `cume_dist`.  It's a PostgreSQL function name.  It should be safe to change it for the string, but not for the function calls, which (for me) would be more confusing rather than less.  I'll put a comment, and we can change the variable in a future PR if we still want to.